actors/builtin/market/testing.go

```diff
@@ -0,0 +1,229 @@
+		return nil, acc, err
+	}
+
+	// next id should be higher than any existing deal
+	acc.Require(int64(st.NextID) > maxDealID, "next id, %d, is not greater than highest id in proposals, %d", st.NextID, maxDealID)
+
+	//
+	// Deal States
+	//
+
+	dealStateCount := uint64(0)
+	dealStates, err := adt.AsArray(store, st.States)
+	if err != nil {
+		return nil, acc, err
+	}
+	var dealState DealState
+	err = dealStates.ForEach(&dealState, func(dealID int64) error {
+		acc.Require(
+			dealState.SectorStartEpoch >= 0,
+			"deal state start epoch undefined %d: %v", dealID, dealState)
+
+		acc.Require(
+			dealState.LastUpdatedEpoch == epochUndefined || dealState.LastUpdatedEpoch >= dealState.SectorStartEpoch,
+			"deal state last updated before sector start %d: %v", dealID, dealState)
+
+		acc.Require(
+			dealState.SlashEpoch == epochUndefined || dealState.SlashEpoch >= dealState.SectorStartEpoch,
+			"deal state slashed before sector start %d: %v", dealID, dealState)
+
+		_, found := allIDs[abi.DealID(dealID)]
+		acc.Require(found, "deal state references deal %d not found in proposals", dealID)
+
+		dealStateCount++
+		return nil
+	})
+	if err != nil {
+		return nil, acc, err
+	}
+
+	//
+	// Pending Proposals
+	//
+
+	pendingProposalCount := uint64(0)
+	pendingProposals, err := adt.AsMap(store, st.PendingProposals)
+	if err != nil {
+		return nil, nil, err
+	}
+	var pendingProposal DealProposal
+	err = pendingProposals.ForEach(&pendingProposal, func(key string) error {
+		proposalCID, err := cid.Parse([]byte(key))
+		if err != nil {
+			return err
+		}
+
+		pcid, err := pendingProposal.Cid()
+		if err != nil {
+			return err
+		}
+		acc.Require(pcid.Equals(proposalCID), "pending proposal's key does not match its CID %v != %v", pcid, proposalCID)
+
+		_, found := proposalCids[pcid]
+		acc.Require(found, "pending proposal with cid %v not fond within proposals %v", pcid, pendingProposals)
+
+		pendingProposalCount++
+		return nil
+	})
+	if err != nil {
+		return nil, acc, err
+	}
+
+	//
+	// Escrow Table and Locked Table
+	//
+
+	escrowTable, err := adt.AsBalanceTable(store, st.EscrowTable)
+	if err != nil {
+		return nil, acc, err
+	}
+
+	lockTableCount := uint64(0)
+	lockTable, err := adt.AsBalanceTable(store, st.LockedTable)
+	if err != nil {
+		return nil, acc, err
+	}
+	var lockedAmount abi.TokenAmount
+	lockedTotal := abi.NewTokenAmount(0)
+	err = (*adt.Map)(lockTable).ForEach(&lockedAmount, func(key string) error {
+		addr, err := address.NewFromBytes([]byte(key))
+		if err != nil {
+			return err
+		}
+		lockedTotal = big.Add(lockedTotal, lockedAmount)
+
+		// every entry in locked table should have a corresponding entry in escrow table that is at least as high
+		escrowAmount, err := escrowTable.Get(addr)
+		if err != nil {
+			return err
+		}
+		acc.Require(escrowAmount.GreaterThanEqual(lockedAmount),
+			"locked funds for %s, %s, greater than escrow amount, %s", addr, lockedAmount, escrowAmount)
+
+		lockTableCount++
+		return nil
```

Comment:
Can we accumulate the per-client/provider collateral amounts from the deals and require the locked amounts here to be at least as large (should exactly match for provider, but client also locks up storage fee, which will be partially paid out and too complicated to recompute the expected value)?